logstash-core/lib/logstash/agent.rb

```diff
@@ -575,6 +577,10 @@ def initialize_geoip_database_metrics(metric)
         dl.gauge(:last_checked_at, nil)
         dl.gauge(:status, nil)
       end
+
+      database_metric = LogStash::Filters::Geoip::DatabaseMetric.new(geoip_metric)
+      database_manager = LogStash::Filters::Geoip::DatabaseManager.instance
+      database_manager.database_metric = database_metric
```

Comment:
Injecting a metric into a singleton seems a little suspicious to me, even if it is being done from the agent.

Is this because the plugins need access to the database manager, and therefore require it to be a singleton because they don't have access to the agent in which they are running?

Reply:
It is sadly due to the tight coupling of DatabaseManager and the geoip plugin. For the backward compatibility, I am a bit hesitant to pass new parameter in [geoip.rb](https://github.com/logstash-plugins/logstash-filter-geoip/blob/master/lib/logstash/filters/geoip.rb#L181) 
Without backward compatibility concerns, thing can work in the following way.
in agent.rb, initialize `@geoip_metric = metric.namespace([:geoip_download_manager])...`, no DatabaseManager singleton
in geoip.rb 
take the metric from agent `geoip_metric = execution_context.agent.geoip_metric`, pass it to DatabaseManager, initialize DatabaseMetric in setup, no more weird assignment in agent.

Regarding the reason for being singleton, multiple geoip plugins need a single scheduler to update database and this scheduler should only be trigger when user config `geoip { }`

Reply:
whenever I change the contract between plugin and DatabaseManager, I need to add constraint 
`gem.add_development_dependency 'logstash-filter-geoip', '>= 7.2.1'` in logstash-core
`s.add_runtime_dependency "logstash-core", ">= 7.14.0"` in logstash-filter-geoip

Reply:
The constraints in the plugin seem necessary (and also painful to manage). I'm with you in being hesitant to change the signature of `DatabaseManager `, but even less excited about hooking `DatabaseManager#subscribe_database_path` to inject the agent.

For now, I think the changes above to have the agent inject it into the singleton instance make the most sense.

But I would prefer the implementation of `DatabaseManager` be made "safe" when there is no agent, so that we don't have to emulate this injection in each and every test, and for the agent to clean the singleton `DatabaseManager` up after itself when being shut down so that we reduce the risk of crossing wires in tests since they instantiate multiple agent instances consecutively in a single process (maybe this is a global `after(:each) { LogStash::Filters::Geoip::DatabaseManager.instance.database_metric = nil }` to unset infected state?)

Reply:
with `let(:db_manager) { manager = Class.new(LogStash::Filters::Geoip::DatabaseManager).instance }`, I should be immune to cross state infection in tests. Let me know if my understanding is wrong.

Reply:
The trouble I was having is that a whole bunch of specs grab that shared reference, then _reach in and mutate it in their own ways often by setting instance variables directly_. This makes it very difficult to ensure that each spec is actually validating the real behaviour, and not some artifact of a prior test's setup.

That said, a lot of that is an inherited artifact of the initial implementation and not directly relevant to the changes being proposed here. I'm okay with paring that off for a separate clean-up effort.